File: Algorithm/264. Ugly Number II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-07-11 09:41:06
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-11 09:41:57
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        uglies = [1]
        idx, primes = [0, 0, 0], [2, 3, 5]
        while len(uglies) < n:
            tmp = min(primes[0]*uglies[idx[0]], primes[1]*uglies[idx[1]], primes[2]*uglies[idx[2]])
            if tmp > uglies[-1]:
                uglies.append(tmp)
            for i in xrange(3):
                if primes[i] * uglies[idx[i]] == tmp:
                    idx[i] += 1
        return uglies[-1]

File: Algorithm/Java/264. Ugly Number II.java
/*
* @Author: WuLC
* @Date:   2017-02-28 13:38:38
* @Last Modified by:   WuLC
* @Last Modified time: 2017-02-28 13:39:13
* @Email: liangchaowu5@gmail.com
*/


// referer: https://discuss.leetcode.com/topic/21882/my-16ms-c-dp-solution-with-short-explanation
public class Solution
{
    public int nthUglyNumber(int n) 
    {
        int[] idices = {0,0,0};
        int[] primes = {2,3,5};
        ArrayList<Integer> uglies = new ArrayList<Integer>();
        uglies.add(1);
        while (uglies.size() < n)
        {
            int tmp = Math.min(uglies.get(idices[0])*primes[0], Math.min(uglies.get(idices[1])*primes[1], uglies.get(idices[2])*primes[2]));
            if(uglies.get(uglies.size()-1) < tmp) uglies.add(tmp);
            for(int i=0; i<3; i++)
            {
                if( uglies.get(idices[i])*primes[i] == tmp ) 
                    idices[i]++;
                    
            }
        }
        return uglies.get(n-1);
    }
}